### PR TITLE
Embedding Op

### LINKED ISSUE #1057: Update Embedding operator

### PR DESCRIPTION
**Pull request by reyna-abhyankar**
**Description of changes:**

Ignore branch name

**Related Issues:**

Linked Issues:
- Issue #

Issues closed by this PR:
- Closes #1057

- - -
This change is [Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1179)

### CODE REVIEW

File: lib/runtime/src/ops/element_unary.cc [hunk 1]
```diff
@@ -6,665 +6,226 @@
...
-Tensor FFModel::elu(const Tensor x, bool inplace, char const *name) {
-  // Currently assume inplace is false
-  assert(!inplace);
-  return this->unary(OP_ELU, x, inplace, name);
+  DeviceSpecific<ElementUnaryPerDeviceState> per_device_state =
+      acc.create_device_specific<ElementUnaryPerDeviceState>(
+          init_kernel(handle,
+                      {input_shape.dims},
+                      {output_shape.dims},
+                      input_shape.data_type));
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/element_unary.cc:70]: @lambda7xx the kernel takes `ArrayShape` and we have `ParallelTensorShape`

**Comment by lambda7xx** [comment]
*[`lib/kernels/src/cuda/layer_norm_kernels.cu` line 36 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLmzMj7NWzMGpxQ1xU:-NgLmzMj7NWzMGpxQ1xV:b-ic1fcb) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/kernels/src/cuda/layer_norm_kernels.cu#L36)):*
> ```Cuda
>   checkCUDA(cudaMalloc(&rstd, sizeof(float) * batch_size));
>   checkCUDA(cudaMalloc(&ds, sizeof(float) * batch_size));
>   checkCUDA(cudaMalloc(&db, sizeof(float) * batch_size));
> ```

how about use `Allocator` to allocate memory? and

**Comment by lambda7xx** [comment]
*[`lib/runtime/src/ops/element_unary.cc` line 70 at r2](https://reviewable.io/reviews/flexflow/FlexFlow/1179#gh-1349747064:-NgLoMon4XYGyTDcVOvv:b-2pudi4) ([raw file](https://github.com/flexflow/FlexFlow/blob/a909c4928b8138eae7118e8ae030172150fb4274/lib/runtime/src/ops/element_unary.cc#L70)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

@lambda7xx the kernel takes `ArrayShape` and we have `ParallelTensorShape`
</blockquote></details>

I think we can use `input` to get its `ArrayShape`

**Comment by lambda7xx** [comment]
*[`lib/runtime/src/ops/element_unary.cc` line 70 at r2](https://reviewable.io/reviews/flexflow/FlexFlow/1179#gh-1349747064:-NgLqUap4oEJmUDsKj0U:bm4ae2l) ([raw file](https://github.com/flexflow/FlexFlow/blob/a909c4928b8138eae7118e8ae030172150fb4274/lib/runtime/src/ops/element_unary.cc#L70)):*
<details><summary><i>Previously, lambda7xx (Lambda(Xiaoxiang) Shi ) wrote…</i></summary><blockquote>

I think we can use `input` to get its `ArrayShape`
</blockquote></details>

forget to pass the op_type. The definition of  init_kernel is below.

```
ElementUnaryPerDeviceState init_kernel(PerDeviceFFHandle const &handle,
                                       ArrayShape const &input_shape,
                                       ArrayShape const &output_shape,
                                       OperatorType op_type,
                                       DataType data_type)
```

**Comment by lambda7xx** [comment]
*[`lib/runtime/src/ops/element_unary.cc` line 216 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLqr3UD0a-30uBnRBg:-NgLqr3UD0a-30uBnRBh:b-xlxmwg) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/element_unary.cc#L216)):*
> ```cpp
>   SimTaskBinding init_binding;
>   init_binding.bind_arg(HANDLE, ff_handle());
>   init_binding.bind_arg(ATTRS, attrs);
> ```

can we bind `ElementScalarUnaryAttrs const &attrs` , then we can get `ElementUnaryPerDeviceState`?

The `ElementScalarUnaryAttrs const &attrs` and ` ElementUnaryAttrs const &attrs` are different class.

**Review by lambda7xx — COMMENTED** [review]
> Reviewed 1 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1179)* status: 1 of 13 files reviewed, 3 unresolved discussions (waiting on @lockshaw, @reyna-abhyankar, and @wmdi)

**Comment by lambda7xx** [comment]
*[`lib/runtime/src/ops/embedding.h` line 20 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLrFiB0kYQO8W6m4j1:-NgLrFiB0kYQO8W6m4j2:b-lsx9ri) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/embedding.h#L20)):*
> ```objc
> CostMetrics measure_operator_cost(SimEnvFactory const &sim_factory,
>                                   EmbeddingAttrs const &attrs,
>                                   InputParallelTensorDesc const &input_shape,
> ```

why ` InputParallelTensorDesc ` ? original code is ` ParallelTensorShape `

**Comment by lambda7xx** [comment]
*[`lib/runtime/src/ops/embedding.cc` line 71 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLrzi85AWMQZOLGVOF:-NgLrzi85AWMQZOLGVOG:b-7ajsuo) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/embedding.cc#L71)):*
> ```cpp
>                  input.shape.get_dim(),
>                  output.shape.get_dim(),
>                  input.shape[legion_dim_t(1)]);
> ```

the original code 

```
    int out_dim = output.domain.hi()[0] - output.domain.lo()[0] + 1;
    int effective_batch_size = output.domain.get_volume() / out_dim;


``
so I think the batch_size should be

```
 int out_dim = output.shape.at(ff_dim_t{0}) + 1;
  int batch_size = output.shape.get_volume() / out_dim;

``

**Comment by lambda7xx** [comment]
*[`lib/runtime/src/ops/embedding.cc` line 71 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLrzi85AWMQZOLGVOF:-NgLsb0rCL-PavH_bLoD:b8cssos) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/embedding.cc#L71)):*
<details><summary><i>Previously, lambda7xx (Lambda(Xiaoxiang) Shi ) wrote…</i></summary><blockquote>

the original code 

```
    int out_dim = output.domain.hi()[0] - output.domain.lo()[0] + 1;
    int effective_batch_size = output.domain.get_volume() / out_dim;


``
so I think the batch_size should be

```
 int out_dim = output.shape.at(ff_dim_t{0}) + 1;
  int batch_size = output.shape.get_volume() / out_dim;

``
</blockquote></details>

the original code

```
    int out_dim = output.domain.hi()[0] - output.domain.lo()[0] + 1;
    int effective_batch_size = output.domain.get_volume() / out_dim;


```
so I think the batch_size should be

```

int out\_dim = output.shape.at(ff\_dim\_t{0}) + 1;  
int batch\_size = output.shape.get\_volume() / out\_dim;

```

**Comment by lambda7xx** [comment]
*[`lib/runtime/src/ops/embedding.cc` line 85 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLsy1y4XFpKoCdnELN:-NgLsy1z3Q6gSE19e7Qe:bd22p0l) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/embedding.cc#L85)):*
> ```cpp
>   auto input = acc.get_tensor<Permissions::RO>(INPUT);
>   auto output = acc.get_tensor<Permissions::RO>(OUTPUT);
>   auto weight_grad = acc.get_tensor_grad<Permissions::RO>(WEIGHT);
> ```

  auto weight_grad = acc.get_tensor_grad<Permissions::RW>(WEIGHT);

**Comment by lambda7xx** [comment]
*[`lib/runtime/src/ops/embedding.cc` line 101 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLt43qDhA5x9zXFdNA:-NgLt43qDhA5x9zXFdNB:bb8gxee) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/embedding.cc#L101)):*
> ```cpp
>                  input.shape.get_dim(),
>                  output.shape.get_dim(),
>                  input.shape[legion_dim_t(1)]);
> ```

the   `input.shape[legion_dim_t(1)]) ` is batch_size? I don't think so.

**Comment by lambda7xx** [comment]
*[`lib/kernels/src/cuda/element_unary_kernels.cu` line 78 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLtSxz26BsvqHr5prZ:-NgLtSxz26BsvqHr5pr_:b-tmkgsc) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/kernels/src/cuda/element_unary_kernels.cu#L78)):*
> ```Cuda
> 
>   ElementUnaryPerDeviceState per_device_state = {
>       handle, inputTensor, outputTensor, actiDesc, op_type, data_type, scalar};
> ```

where is scalar

**Comment by lambda7xx** [comment]
*[`lib/kernels/src/cuda/layer_norm_kernels.cu` line 36 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLmzMj7NWzMGpxQ1xU:-NgLthHS5rr08JtsIRJh:b-saksvv) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/kernels/src/cuda/layer_norm_kernels.cu#L36)):*
<details><summary><i>Previously, lambda7xx (Lambda(Xiaoxiang) Shi ) wrote…</i></summary><blockquote>

how about use `Allocator` to allocate memory? and
</blockquote></details>

I set a  pr about layernorm [LayerNorm OP draft by lambda7xx · Pull Request #1186 · flexflow/FlexFlow (github.com)](https://github.com/flexflow/FlexFlow/pull/1186)

**Review by reyna-abhyankar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1179)* status: 1 of 13 files reviewed, 7 unresolved discussions (waiting on @lambda7xx, @lockshaw, and @wmdi)

___
*[`lib/kernels/src/cuda/element_unary_kernels.cu` line 78 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLtSxz26BsvqHr5prZ:-NgPi2jt013xKbjiLdy7:b-qpz8j) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/kernels/src/cuda/element_unary_kernels.cu#L78)):*
<details><summary><i>Previously, lambda7xx (Lambda(Xiaoxiang) Shi ) wrote…</i></summary><blockquote>

where is scalar
</blockquote></details>

Done. See other comment (I think we should merge the two attrs classes)
___
*[`lib/kernels/src/cuda/layer_norm_kernels.cu` line 36 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLmzMj7NWzMGpxQ1xU:-NgPczaR3uoMOAgsN9kB:b-w6ubeq) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/kernels/src/cuda/layer_norm_kernels.cu#L36)):*
<details><summary><i>Previously, lambda7xx (Lambda(Xiaoxiang) Shi ) wrote…</i></summary><blockquote>

I set a  pr about layernorm [LayerNorm OP draft by lambda7xx · Pull Request #1186 · flexflow/FlexFlow (github.com)](https://github.com/flexflow/FlexFlow/pull/1186)
</blockquote></details>

Ok, we can use your layer norm PR
I think you can do something like this:



_Code snippet:_
```Cuda
int n = 6; // number of pointers
mean = (float *) allocator.allocate(sizeof(float) * batch_size * n);
rstd = (float *) mean + batch_size;
...
```
___
*[`lib/runtime/src/ops/element_unary.cc` line 216 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLqr3UD0a-30uBnRBg:-NgPgLmL3ZjuT6t4DN7h:bhbynpc) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/element_unary.cc#L216)):*
<details><summary><i>Previously, lambda7xx (Lambda(Xiaoxiang) Shi ) wrote…</i></summary><blockquote>

can we bind `ElementScalarUnaryAttrs const &attrs` , then we can get `ElementUnaryPerDeviceState`?

The `ElementScalarUnaryAttrs const &attrs` and ` ElementUnaryAttrs const &attrs` are different class.
</blockquote></details>

I actually think we should merge them. @lockshaw the op type will determine what is executed in the kernel anyway
___
*[`lib/runtime/src/ops/embedding.h` line 20 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLrFiB0kYQO8W6m4j1:-NgPhFiECDTcPb_jcDeq:b-urxsqh) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/embedding.h#L20)):*
<details><summary><i>Previously, lambda7xx (Lambda(Xiaoxiang) Shi ) wrote…</i></summary><blockquote>

why ` InputParallelTensorDesc ` ? original code is ` ParallelTensorShape `
</blockquote></details>

`InputParallelTensorDesc` tells us if an input is trainable or not. Useful for the binding
___
*[`lib/runtime/src/ops/embedding.cc` line 71 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLrzi85AWMQZOLGVOF:-NgPhtAb0jw3SUpZoiiX:b-xj8s5t) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/embedding.cc#L71)):*
<details><summary><i>Previously, lambda7xx (Lambda(Xiaoxiang) Shi ) wrote…</i></summary><blockquote>

the original code

```
    int out_dim = output.domain.hi()[0] - output.domain.lo()[0] + 1;
    int effective_batch_size = output.domain.get_volume() / out_dim;


```
so I think the batch_size should be

```

int out\_dim = output.shape.at(ff\_dim\_t{0}) + 1;  
int batch\_size = output.shape.get\_volume() / out\_dim;

```
</blockquote></details>

@lockshaw 
___
*[`lib/runtime/src/ops/embedding.cc` line 85 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLsy1y4XFpKoCdnELN:-NgPhynj0JtXrfSTX_eS:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/embedding.cc#L85)):*
<details><summary><i>Previously, lambda7xx (Lambda(Xiaoxiang) Shi ) wrote…</i></summary><blockquote>

auto weight_grad = acc.get_tensor_grad<Permissions::RW>(WEIGHT);
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/embedding.cc` line 101 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLt43qDhA5x9zXFdNA:-NgPhzzV1k6AHZ0Gjy9A:b5d84gi) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/embedding.cc#L101)):*
<details><summary><i>Previously, lambda7xx (Lambda(Xiaoxiang) Shi ) wrote…</i></summary><blockquote>

the   `input.shape[legion_dim_t(1)]) ` is batch_size? I don't think so.
</blockquote></details>

Done. See previous comment.

**Review by lockshaw — CHANGES_REQUESTED** [review]
> Reviewed 4 of 5 files at r1, 3 of 4 files at r2, 2 of 2 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1179)* status: all files reviewed, 13 unresolved discussions (waiting on @lambda7xx, @reyna-abhyankar, and @wmdi)

___
*[`lib/kernels/include/kernels/element_unary_kernels.h` line 19 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgaCCZ00TeMUt8Zjwdw:-NgaCCZ00TeMUt8Zjwdx:b-cuvpne) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/kernels/include/kernels/element_unary_kernels.h#L19)):*
> ```objc
>   OperatorType op_type;
>   DataType data_type;
>   float scalar;
> ```

Does this really compile without the `req`?



_Suggestion:_
```objc
  req<float> scalar;
```
___
*[`lib/kernels/include/kernels/layer_norm_kernels.h` line 38 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgaCTku2gLDwQl1J5n9:-NgaCTku2gLDwQl1J5nA:b-eqa5ca) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/kernels/include/kernels/layer_norm_kernels.h#L38)):*
> ```objc
>                      GenericTensorAccessorW const &beta_grad,
>                      DataType data_type,
>                      int64_t batch_size,
> ```

Isn't this part of the shape so it can just be accessed via `input`?
___
*[`lib/kernels/include/kernels/layer_norm_kernels.h` line 39 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgaCY9hAmRGzYKs5mmv:-NgaCY9hAmRGzYKs5mmw:b-n1vqjk) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/kernels/include/kernels/layer_norm_kernels.h#L39)):*
> ```objc
>                      DataType data_type,
>                      int64_t batch_size,
>                      int64_t num_elements,
> ```

Isn't this part of the shape so it can jut be accessed via one of the weights?
___
*[`lib/kernels/src/cuda/element_unary_kernels.cu` line 78 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLtSxz26BsvqHr5prZ:-NgaCl-uApP3TNfooseO:b-97fq0k) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/kernels/src/cuda/element_unary_kernels.cu#L78)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

Done. See other comment (I think we should merge the two attrs classes)
</blockquote></details>

Are you still in favor of this after the meeting yesterday? I'd like to keep the `attrs` separate, though I don't really care what happens with them in `kernels`
___
*[`lib/kernels/src/cuda/layer_norm_kernels.cu` line 36 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLmzMj7NWzMGpxQ1xU:-NgaD4y4B1562Yoglcny:b-b8npb7) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/kernels/src/cuda/layer_norm_kernels.cu#L36)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

Ok, we can use your layer norm PR
I think you can do something like this:
</blockquote></details>

Yeah `cudaMalloc` should be replaced by `Allocator`
___
*[`lib/kernels/src/hip/layer_norm_kernels.cpp` line 33 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgaDKKbAO2aCCv3LAYJ:-NgaDKKbAO2aCCv3LAYK:b-3p2zfr) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/kernels/src/hip/layer_norm_kernels.cpp#L33)):*
> ```cpp
>                                     int64_t effective_batch_size) {
>   float *mean, *rstd, *ds, *db, *scale, *bias;
>   checkCUDA(cudaMalloc(&mean, sizeof(float) * batch_size));
> ```

Use `Allocator` here
___
*[`lib/kernels/src/hip/layer_norm_kernels.cpp` line 191 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgaDS402S4atp49On6h:-NgaDS402S4atp49On6i:bjhf0oc) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/kernels/src/hip/layer_norm_kernels.cpp#L191)):*
> ```cpp
>                        GenericTensorAccessorW const &beta_grad,
>                        DataType data_type,
>                        int64_t batch_size,
> ```

Isn't this accessible through the tensor shapes?
___
*[`lib/runtime/src/ops/element_unary.cc` line 178 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgaDzQP7bBBM674laOr:-NgaDzQP7bBBM674laOs:undefined) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/element_unary.cc#L178)):*
> ```cpp
>   init_binding.bind_arg(HANDLE, ff_handle());
>   init_binding.bind_arg(ATTRS, attrs);
>   init_binding.bind_arg(INPUT_SHAPE, input_parallel_tensor_shape(0));
> ```





_Suggestion:_
```cpp
init_binding.bind_arg(INPUT_SHAPE, input_shape);
```
___
*[`lib/runtime/src/ops/element_unary.cc` line 216 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLqr3UD0a-30uBnRBg:-NgaEAgO2dL-W9RNejDr:b-lxan63) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/element_unary.cc#L216)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

I actually think we should merge them. @lockshaw the op type will determine what is executed in the kernel anyway
</blockquote></details>

I'd like to keep them separate at the `op-attrs` level, but I don't care what happens to them them in the `runtime/ops` and `kernels` levels
___
*[`lib/runtime/src/ops/embedding.cc` line 71 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLrzi85AWMQZOLGVOF:-NgaEdbX1KeA0RVZEcvO:b9c000g) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/embedding.cc#L71)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

@lockshaw
</blockquote></details>

Should be `input.shape[ff_dim_t(0)]` I think, as this is just a `TensorShape` and not a `ParallelTensorShape` and so there shouldn't be a parallel dimension present
___
*[`lib/runtime/src/ops/embedding.cc` line 85 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLsy1y4XFpKoCdnELN:-NgaEyRL3xOSTjRciZth:bq85doa) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/embedding.cc#L85)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I'm still seeing `Permissions::RO` here...
___
*[`lib/runtime/src/ops/embedding.cc` line 101 at r4](https://reviewable.io/reviews/flexflow/FlexFlow/1179#-NgLt43qDhA5x9zXFdNA:-NgaF6fQ5ODIFJ-ji2ww:b-49b4m0) ([raw file](https://github.com/flexflow/FlexFlow/blob/e1fb8a8ff8378701252a30002d0373de19ef2c82/lib/runtime/src/ops/embedding.cc#L101)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

Done. See previous comment.
</blockquote></details>

Should be `ff_dim_t(0)`

**Pull request closed without merging**